py/gc: Reduce code size.

aykevl opened the pull request:
I've found a way to reduce the code size slightly.

The reason it is smaller is because the stack pointer check in `gc_collect_root()` is unnecessary: the stack will always be empty when entering `gc_collect_root()`, so the check for an overflow is superfluous.

Code size reductions from when I tested this (a few weeks ago):
```
unix:            -48
minimal CROSS=1: -32
stm32:           -32
```

aykevl committed:
py/gc: Reduce code size.

unix:            -48
minimal CROSS=1: -32
stm32:           -32

dpgeorge commented on py/gc.c:
```diff
@@ -236,7 +218,21 @@ STATIC void gc_drain_stack(void) {
-            VERIFY_MARK_AND_PUSH(ptr);
+            if (VERIFY_PTR(ptr)) {
+                // Mark and push this pointer
+                // TODO: almost-duplicate code in gc_collect_root
```
I don't think these TODO's should stay because there's nothing "to do": it's not really possible to refactor this (or part of this) code into a common macro.  I'd be happy not to have any comment about the almost-duplicated code.

aykevl replied:
Fixed.

dpgeorge commented on py/gc.c:
```diff
@@ -337,8 +333,18 @@ void gc_collect_start(void) {
+                ATB_HEAD_TO_MARK(block);
+                *MP_STATE_MEM(gc_sp)++ = block;
+            }
+            gc_drain_stack();
```
This can probably move up 1 line into the if-statement.

aykevl replied:
For some reason, moving it up increases code size by 4 bytes in gcc 5.4.1 (for the minimal port).

dpgeorge replied:
IMO that's acceptable: it's ok to have a tiny increase in size if the code is logically better.  In this case you would expect the code size to be the same if the call is inside or outside the if, and maybe a different gcc (or non-gcc compiler) does something slightly different and makes the code smaller or the same.  And remember that gc.c is actually compiled with -O3 by default (grep for CSUPEROPT) so size optimisations are not always what they seem :)

Is summary, I'd prefer if this line were within the if to make the logic clearer.

aykevl replied:
Done.

dpgeorge reviewed (commented):
Looks good!

Then this may open up a further (code + speed) optimisation: gc_drain_stack() is always called right after pushing a new block so gc_drain_stack() could do this pushing itself (or use the value right away).  If you attempt this optimisation, please put it in a separate commit (but same PR is ok).

aykevl commented:
I've played around a bit with the code per your suggestion and I think I found some more optimisations that can reduce code size by ~72 bytes. But these are a bit more involved so I want to do those in a separate PR to not block this PR.

aykevl committed:
py/gc: Pass first block to gc_drain_stack().

This saves a bit in code size:
minimal CROSS=1: -44
unix:            -96

aykevl committed:
py/gc: Make stack pointer a local variable.

This saves a bit in code size and .bss RAM size.

                 .text  .bss
minimal CROSS=1: -28    -4
unix (64-bit):   -64    -8

dpgeorge commented:
> I've played around a bit with the code per your suggestion and I think I found some more optimisations that can reduce code size by ~72 bytes (minimal CROSS=1). But these are a bit more involved so I want to do those in a separate PR to not block this PR.

OK!

aykevl commented:
I have pushed the second commit anyway, as it is strongly related (and would move `gc_drain_stack` up to within the if condition anyway). Now the size calculations are slightly off per commit but will be correct when combined.

The last commit I have ready moves `gc_sp` into a local variable, saving 28 bytes code and 4 bytes .bss on thumb2. I can push this commit if preferred (to review them all in one go).

aykevl commented on py/gc.c:
```diff
@@ -203,29 +203,13 @@ bool gc_is_locked(void) {
+// children: mark the unmarked child blocks and put those newly marked
+// blocks on the stack. When all children have been checked, pop off the
+// topmost block on the stack and repeat with that one.
+STATIC void gc_drain_stack(size_t block) {
```
Maybe this function needs to be renamed - it isn't really draining the stack anymore as no one outside `gc_drain_stack` pushes anything to the stack.
Maybe something like `gc_mark_children_recursive`? It isn't recursive in the common way (calling itself) but it implements a sort of recursion by using it's own stack, hence the name.

dpgeorge replied:
Yes it makes sense to rename it.  But it would be good to stay away from the word "recursive" if possible because embedded stuff is ideally not recursive :)  Something like `gc_mark_subtree` or `gc_mark_descendants`.

aykevl replied:
I've renamed it to `gc_mark_subtree`.

dpgeorge commented on py/gc.c:
```diff
@@ -203,29 +203,13 @@ bool gc_is_locked(void) {
+// topmost block on the stack and repeat with that one.
+STATIC void gc_drain_stack(size_t block) {
+    // Start with the block passed in the argument.
+    while (1) {
```
I know this is picky and not documented in the conventions, but the codebase styling does slightly prefer `for (;;) {... }`  as opposed to `while (1) { ... }`.

aykevl replied:
Fixed.

dpgeorge commented:
> I have pushed the second commit anyway, as it is strongly related (and would move gc_drain_stack up to within the if condition anyway). Now the size calculations are slightly off per commit but will be correct when combined.

Thanks, the change looks good.  Computing size change at my end I get, for the first commit:
```
   bare-arm:    +0 
minimal x86:  -112 
   unix x64:  -144 
unix nanbox:  -112 
      stm32:   -56 
     cc3200:   -48 
    esp8266:   -28 
      esp32:   -20 
```
and the second commit:
```
   bare-arm:    +0 
minimal x86:   -64 
   unix x64:   -80 
unix nanbox:   -48 
      stm32:   -44 
     cc3200:   -48 
    esp8266:   -20 
      esp32:   -28 
```
That's great!

In general one should also be wary of change in performance, data/bss usage, and C stack usage.  But I think all of those are OK here.

> The last commit I have ready moves gc_sp into a local variable, saving 28 bytes code and 4 bytes .bss on thumb2. I can push this commit if preferred (to review them all in one go).

It sounds like this commit is closely related to the ones here, so please go ahead and push it if it's ready.  Then they can all go in together.

aykevl commented:
I have pushed the last commit as well. It is a relatively small change.

As it is now possible I've tried to make `gc_stack` a local variable but it turns out this increases code size, which I didn't expect.

dpgeorge commented:
> As it is now possible I've tried to make gc_stack a local variable but it turns out this does not bring a big reduction in code size 

It's still a decent reduction in code size, I measure (for the 3rd commit here):
```
minimal x86:   -36 [incl -4(bss)]
   unix x64:   -56 [incl -8(bss)]
unix nanbox:   -36 [incl -4(bss)]
      stm32:   -36 [incl -4(bss)]
     cc3200:   -32 
    esp8266:   -36 
      esp32:   -28 
```
Code size reduction is good, but to remove even 4 bytes from the bss (the global uPy state) is really great.  Memory is very precious and in this case I don't see any downside from the 3rd commit, because it makes a global variable local and that should improve speed as well.

> and seems risky to me.

Can you elaborate?  Apart from thinking through any consequences of the patch to multithreading (and I don't think there'll be any issues here) I don't see any other issues.

aykevl commented:
> As it is now possible I've tried to make gc_stack a local variable but it turns out this does not bring a big reduction in code size and seems risky to me.

In this last sentence I was talking about a hypothetical next step here (making `gc_stack` local), not the last of the 3 commits (making `gc_sp` local). Sorry for the confusion.

dpgeorge commented:
> In this last sentence I was talking about a hypothetical next step here (making gc_stack local),

Ah, yes, I see, sorry I didn't read it correctly.  I agree it's a bit risky because gc_stack is quite large (eg 256 bytes) and may overflow the C stack.

dpgeorge commented:
Thanks a lot for this, it's a nice optimisation!

Rebased, commit messages slightly reworded, and merged in ea7cf2b738c3b103b472547500fb0107b956fc0c through 736faef2233fe9c721b940a108b28808d4c7f7a0
